ATHIYA PUTERI HIDAYAT - 24060121140128 - LAB MACHINE LEARNING A1

PERTEMEMUAN 2 - EVALUASI ALGORITMA (IRIS)

In [ ]:
# import libary yang dibutuhkan yaitu pandas disingkat menjadi pd
import pandas as pd

In [ ]:
# Mendefinisikan link dataset
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

# Mendefinisikan list berisi kolom (attribute) dataset
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']

# Membaca dataset dengan library panda
dataset = pd.read_csv(url, names=names)

In [ ]:
dataset.head(5)

sepal-length  sepal-width  petal-length  petal-width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [ ]:
# Mencetak bentuk dataset
dataset.shape

(150, 5)

Training-Test Split

In [ ]:
# import library yang dibutuhkan
from sklearn import model_selection

In [ ]:
# Split-out validation dataset

# Mengkonversi dataset menjadi array
array = dataset.values

# Memisahkan fitur kolom 0 hingga 4 sebagai variabel independen (X)
X = array[:,0:4]

# Memilih kolom 4 sebagai variabel dependen (Y)
Y = array[:,4]

# Menentukan ukuran validasi (20% dari total data)
validation_size = 0.20

# Menentukan seed untuk reproducibility
seed = 7

# Membagi dataset menjadi data latih dan data validasi menggunakan train_test_split
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [ ]:
# Print Split-Out

# Menampilkan shape dari set data X_train
print("X_train :", X_train.shape)

# Menampilkan shape dari set data X_validation
print("X_validation :", X_validation.shape)

# Menampilkan shape dari label data Y_train
print("Y_train :", Y_train.shape)

# Menampilkan shape dari label data Y_validation
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


In [ ]:
# Test options and evaluation metric
seed = 7

scoring = 'accuracy'

In [ ]:
# Import Library berisi algoritma klasifikasi yang akan digunakan

# Mengimport kelas GaussianProcessClassifier dari modul sklearn.gaussian_process
from sklearn.gaussian_process import GaussianProcessClassifier

# Mengimport kelas AdaBoostClassifier dari modul sklearn.ensemble
from sklearn.ensemble import AdaBoostClassifier

# Mengimport kelas DecisionTreeClassifier dari modul sklearn.tree
from sklearn.tree import DecisionTreeClassifier

# Spot Check Algorithms

# Inisialisasi model Gaussian, Ada Boost, dan Decision Tree
models = []

# Menambahkan model Gaussian ke dalam list models
models.append(('Gaussian Process', GaussianProcessClassifier()))

# Menambahkan model Ada Boost ke dalam list models
models.append(('Ada Boost', AdaBoostClassifier()))

# Menambahkan model Decision Tree ke dalam list models
models.append(('Decision Tree', DecisionTreeClassifier()))

In [ ]:
# evaluate each model in turn
results = []

# Menyimpan nama-nama model
names = []

for name, model in models:
  # Membagi data menjadi training dan validasi menggunakan k-fold validation
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # Menghitung Nilai Akurasi
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  # Memasukan hasil nilai akurasi di results
  results.append(cv_results)
  names.append(name)

  # MEngeluarkan pesan yang berisi nama, rata - rata, dan standar deviasi
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

Gaussian Process: 0.983333 (0.033333)
Ada Boost: 0.966667 (0.040825)
Decision Tree: 0.958333 (0.076830)


Dari hasil evaluasi di atas, dapat dilihat bahwa Gaussian memperoleh nilai akurasi perkiraan terbesar yaitu 0.983333. Dengan mengetahui model yang paling tinggi itu Gaussian dapat dilanjutkan dengan melakukan pengujian keakuratan model Gaussian.

In [ ]:
#Import Library yang dibutuhkan
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
# Make predictions on validation dataset
gp = GaussianProcessClassifier()

# Melatih model Gaussian Process menggunakan data latih
gp.fit(X_train, Y_train)

# Menguji model klasisifikasi (Gaussian Process) yang dihasilkan dari proses pelatihan data dengan testing data
predictions = gp.predict(X_validation)

#Mencetak accuracy score menggunakan Y validation (memprediksi masuk kelas mana)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8666666666666667
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



*   Data validasi yang diterapkan dengan model Gaussian memiliki hasil evaluasi atau tingkat akurasi mencapai nilai 0.8666666666666667 atau 86.67%.
*   Confusion matrix akan menunjukan secara detail ketepatan model dalam memprediksi kelas untuk setiap sampel. Hasil menunjukan bahwa model Gaussian memiliki kemampuan yang baik dalam memprediksi tiap kelas. Terdapat 7 contoh berhasil diklasifikasi di kelas Iris-setosa, 12 contoh berhasil diklasifikasi di kelas Iris-versicolor, dan terdapat 11 contoh berhasil diklasifikasi di kelas Iris-virginica.

*   Classification Support meenandakan kemampuan model dalam mengindentifikasi dan mengklasifikasi setiap kelas secara akurat. Pada classification support menunjukan bahwa model memiliki precision tinggi pada Iris-setosa, recall semua mencapai akurasi sebesar 0.88 atau 88%, dan f1-score semua mencapai akurasi sebesar 0.87 atau 87%.


PERTEMEMUAN 2 - EVALUASI ALGORITMA (DIABETIC RETINOPATHY DEBRECEN)

In [1]:
pip install ucimlrepo

In [2]:
from ucimlrepo import fetch_ucirepo

# fetch dataset
diabetic_retinopathy_debrecen = fetch_ucirepo(id=329)

# data (as pandas dataframes)
X = diabetic_retinopathy_debrecen.data.features
y = diabetic_retinopathy_debrecen.data.targets

# metadata
print(diabetic_retinopathy_debrecen.metadata)

# variable information
print(diabetic_retinopathy_debrecen.variables)

{'uci_id': 329, 'name': 'Diabetic Retinopathy Debrecen', 'repository_url': 'https://archive.ics.uci.edu/dataset/329/diabetic+retinopathy+debrecen', 'data_url': 'https://archive.ics.uci.edu/static/public/329/data.csv', 'abstract': 'This dataset contains features extracted from the Messidor image set to predict whether an image contains signs of diabetic retinopathy or not. ', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 1151, 'num_features': 19, 'feature_types': ['Integer', 'Real'], 'demographics': [], 'target_col': ['Class'], 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2014, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C5XP4P', 'creators': ['Balint Antal', 'Andras Hajdu'], 'intro_paper': {'title': 'An ensemble-based system for automatic screening of diabetic retinopathy', 'authors': 'B. Antal, A. Hajdu', 'published_in': 'Knowledge-Based Syste

In [3]:
# import libary yang dibutuhkan yaitu pandas disingkat menjadi pd
import pandas as pd

In [4]:
# Mendefinisikan link dataset
url = "https://archive.ics.uci.edu/static/public/329/data.csv"

# Membaca dataset dari URL
raw_dataset = pd.read_csv(url)

# Memilih kolom-kolom tertentu dari dataset
dataset = raw_dataset.iloc[:, 0:19]

In [5]:
dataset

quality  pre_screening  ma1  ma2  ma3  ma4  ma5  ma6   exudate1  \
0           1              1   22   22   22   19   18   14  49.895756   
1           1              1   24   24   22   18   16   13  57.709936   
2           1              1   62   60   59   54   47   33  55.831441   
3           1              1   55   53   53   50   43   31  40.467228   
4           1              1   44   44   44   41   39   27  18.026254   
...       ...            ...  ...  ...  ...  ...  ...  ...        ...   
1146        1              1   34   34   34   33   31   24   6.071765   
1147        1              1   49   49   49   49   45   37  63.197145   
1148        1              0   49   48   48   45   43   33  30.461898   
1149        1              1   39   36   29   23   13    7  40.525739   
1150        1              1    7    7    7    7    7    5  69.423565   

       exudate2   exudate3  exudate3.1  exudate5  exudate6  exudate7  \
0     17.775994   5.270920    0.771761  0.018632  0.006864  0.003923   
1     23.799994   3.325423    0.234185  0.003903  0.003903  0.003903   
2     27.993933  12.687485    4.852282  1.393889  0.373252  0.041817   
3     18.445954   9.118901    3.079428  0.840261  0.272434  0.007653   
4      8.570709   0.410381    0.000000  0.000000  0.000000  0.000000   
...         ...        ...         ...       ...       ...       ...   
1146   0.937472   0.031145    0.003115  0.000000  0.000000  0.000000   
1147  27.377668   8.067688    0.979548  0.001552  0.000000  0.000000   
1148  13.966980   1.763305    0.137858  0.011221  0.000000  0.000000   
1149  12.604947   4.740919    1.077570  0.563518  0.326860  0.239568   
1150   7.031843   1.750548    0.046597  0.021180  0.008472  0.000000   

      exudate8  macula_opticdisc_distance  opticdisc_diameter  \
0     0.003923                   0.486903            0.100025   
1     0.003903                   0.520908            0.144414   
2     0.007744                   0.530904            0.128548   
3     0.001531                   0.483284            0.114790   
4     0.000000                   0.475935            0.123572   
...        ...                        ...                 ...   
1146  0.000000                   0.537470            0.116795   
1147  0.000000                   0.516733            0.124190   
1148  0.000000                   0.560632            0.129843   
1149  0.174584                   0.485972            0.106690   
1150  0.000000                   0.556192            0.088957   

      am_fm_classification  
0                        1  
1                        0  
2                        0  
3                        0  
4                        0  
...                    ...  
1146                     0  
1147                     0  
1148                     0  
1149                     1  
1150                     0  

[1151 rows x 19 columns]

In [6]:
# Mencetak bentuk dataset
dataset.shape

(1151, 19)

In [7]:
# import library yang dibutuhkan
from sklearn import model_selection

In [45]:
# Split-out validation dataset

# Mengkonversi dataset menjadi array
array = dataset.values

# Memisahkan fitur kolom 0 hingga 19 sebagai variabel independen (X)
X = array[:, 0:19]

# Memilih kolom 18 sebagai variabel dependen (Y)
Y = array[:, 18]

# Menentukan ukuran validasi (20% dari total data)
validation_size = 0.20

# Menentukan seed untuk reproducibility
seed = 7

# Membagi dataset menjadi data latih dan data validasi menggunakan train_test_split
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [46]:
# Print Split-Out

# Menampilkan shape dari set data X_train
print("X_train :", X_train.shape)

# Menampilkan shape dari set data X_validation
print("X_validation :", X_validation.shape)

# Menampilkan shape dari label data Y_train
print("Y_train :", Y_train.shape)

# Menampilkan shape dari label data Y_validation
print("Y_validation :", Y_validation.shape)

X_train : (920, 19)
X_validation : (231, 19)
Y_train : (920,)
Y_validation : (231,)


In [47]:
# Ini adalah penentuan nilai seed untuk reproducibility.
seed = 7

# Mengukur sejauh mana model dapat mengklasifikasikan data dengan benar.
scoring = 'accuracy'

In [48]:
# Import Library berisi algoritma klasifikasi yang akan digunakan

# Mengimport kelas KNeighborsClassifier dari modul sklearn.neighbors
from sklearn.neighbors import KNeighborsClassifier

# Mengimport kelas GaussianNB dari modul sklearn.naive_bayes
from sklearn.naive_bayes import GaussianNB

# Mengimport kelas SVC (Support Vector Classification) dari modul sklearn.svm
from sklearn.svm import SVC

# Spot Check Algorithms
# Inisialisasi model K-Nearest Neighbors (KNN), Naive Bayes (NB), dan Support Vector Machine (SVM)
models = []

# Menambahkan model KNN ke dalam list models
models.append(('KNN', KNeighborsClassifier()))

# Menambahkan model Naive Bayes (NB) ke dalam list models
models.append(('NB', GaussianNB()))

# Menambahkan model Support Vector Machine (SVM) ke dalam list models
models.append(('SVM', SVC()))

In [49]:
# evaluate each model in turn
results = []

# Menyimpan nama-nama model
names = []

for name, model in models:
  # Membagi data menjadi training dan validasi menggunakan k-fold validation
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  # Menghitung Nilai Akurasi
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  # Memasukan hasil nilai akurasi di results
  results.append(cv_results)
  names.append(name)

  # MEngeluarkan pesan yang berisi nama, rata - rata, dan standar deviasi
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.736957 (0.045549)
NB: 1.000000 (0.000000)
SVM: 0.755435 (0.027174)


Dari hasil evaluasi di atas, dapat dilihat bahwa Gaussian Naive Bayes (GaussianNB) memperoleh nilai akurasi perkiraan terbesar yaitu 1.000000. Dengan mengetahui model yang paling tinggi itu Gaussian Naive Bayes (GaussianNB) dapat dilanjutkan dengan melakukan pengujian keakuratan model Gaussian Naive Bayes (GaussianNB).

In [50]:
#Import Library yang dibutuhkan
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [51]:
# Make predictions on validation dataset
nb = GaussianNB()

# Melatih model Gaussian Process menggunakan data latih
nb.fit(X_train, Y_train)

# Menguji model klasisifikasi (Gaussian Process) yang dihasilkan dari proses pelatihan data dengan testing data
predictions = nb.predict(X_validation)

#Mencetak accuracy score menggunakan Y validation (memprediksi masuk kelas mana)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

1.0
[[153   0]
 [  0  78]]
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00       153
         1.0       1.00      1.00      1.00        78

    accuracy                           1.00       231
   macro avg       1.00      1.00      1.00       231
weighted avg       1.00      1.00      1.00       231



*   Validasi dataset yang diterapkan dengan model Gaussian Naive Bayes (GaussianNB) menghasilkan hasil evaluasi yang sempurna yaitu mencapai nilai 1.0 atau 100%. Hak tersebut menunjukan bahwa model Gaussian Naive Bayes (GaussianNB) berhasil mengklasifikasi seluruh data validasi tanpa kesalahan atau kekurangan.
*   Confusion matrix akan menunjukan secara detail ketepatan model dalam memprediksi kelas untuk setiap sampel. Pada hasil confusion matrix menunjukan pada data validasi diklasifikasi dengan sempurna. Terdapat 153 instance yang semuanya diprediksi benar sebagai kelas 0.0. Lalu terdapat 78 instance yang semuanya diprediksi benar sebagai kelas 1.0.

*   Classification Support meenandakan kemampuan model dalam mengindentifikasi dan mengklasifikasi setiap kelas secara akurat karena menunjukan bahwa model memiliki precision, recall, dan f1-score yang sangat tinggi pada tiap kelas.




In [52]:
# Make predictions on validation dataset
knn = KNeighborsClassifier()

# Melatih model Gaussian Process menggunakan data latih
knn.fit(X_train, Y_train)

# Menguji model klasisifikasi (Gaussian Process) yang dihasilkan dari proses pelatihan data dengan testing data
predictions = knn.predict(X_validation)

#Mencetak accuracy score menggunakan Y validation (memprediksi masuk kelas mana)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.7402597402597403
[[118  35]
 [ 25  53]]
              precision    recall  f1-score   support

         0.0       0.83      0.77      0.80       153
         1.0       0.60      0.68      0.64        78

    accuracy                           0.74       231
   macro avg       0.71      0.73      0.72       231
weighted avg       0.75      0.74      0.74       231



In [53]:
# Make predictions on validation dataset
svc = SVC()

# Melatih model Gaussian Process menggunakan data latih
svc.fit(X_train, Y_train)

# Menguji model klasisifikasi (Gaussian Process) yang dihasilkan dari proses pelatihan data dengan testing data
predictions = svc.predict(X_validation)

#Mencetak accuracy score menggunakan Y validation (memprediksi masuk kelas mana)
print(accuracy_score(Y_validation, predictions))

print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.7445887445887446
[[125  28]
 [ 31  47]]
              precision    recall  f1-score   support

         0.0       0.80      0.82      0.81       153
         1.0       0.63      0.60      0.61        78

    accuracy                           0.74       231
   macro avg       0.71      0.71      0.71       231
weighted avg       0.74      0.74      0.74       231

